sources/vault.move

```diff
@@ -0,0 +1,122 @@
+/// User can deposit/withdraw multiple fungible assets to/from different markets
+module playground::vault {
+    use std::signer;
+    use std::string;
+
+    use aptos_std::table::{Self, Table};
+    use aptos_framework::fungible_asset::{Self, FungibleAsset, Metadata};
+    use aptos_framework::object::{Self, Object};
+
+    #[test_only]
+    use aptos_framework::primary_fungible_store;
+    #[test_only]
+    use std::option;
+    use std::vector;
+
+    const ERR_UNAUTHORIZED: u64 = 1;
+    const ERR_INSUFFICIENT_FUNDS: u64 = 2;
+
+    /// Market holds user deposited assets
+    struct Market has key {
+        assets: Object<FungibleAsset>
+    }
+
+    /// Balance keeps track of user assets in different markets
+    struct User has key {
+        markets: Table<Object<Market>, u64>
+    }
+
+    public fun init_market(admin: &signer, asset: Object<Metadata>): Object<Market> {
+        assert!(signer::address_of(admin) == @playground, ERR_UNAUTHORIZED);
+
+        let asset_name = fungible_asset::name(asset);
+
+        let fa_seed = vector::empty<u8>();
+        vector::append(&mut fa_seed, b"fa-");
+        vector::append(&mut fa_seed, *string::bytes(&asset_name));
+        let (fa_signer, fa_obj) = create_object<FungibleAsset>(admin, fa_seed);
+        move_to(&fa_signer, fungible_asset::zero(asset));
+
+        let market_seed = vector::empty<u8>();
+        vector::append(&mut market_seed, b"market-");
+        vector::append(&mut market_seed, *string::bytes(&asset_name));
+        let (market_signer, market_obj) = create_object<Market>(admin, *string::bytes(&asset_name));
+        move_to(&market_signer, Market {
+            assets: fa_obj
+        });
+
+        market_obj
+    }
+
+    /// Deposit fungible asset to a market
+    public fun deposit(account: &signer, market_obj: Object<Market>, fa: FungibleAsset) acquires Market, User {
+        let amount = fungible_asset::amount(&fa);
+
+        let market = borrow_global_mut<Market>(object::object_address(&market_obj));
+        fungible_asset::merge(&mut market.assets, fa);
```

Comment:
Don't merge like this. You need to do fungible_asset::deposit(market_object_address, fa) where market_object_address is that single object I mentioned earlier that has both FungibleAsset (it's a store) and Market resource